File: lib/model/result.py
from sqlalchemy import (Column,
                        ForeignKey,
                        Integer,
                        String,
                        UniqueConstraint)
from sqlalchemy.orm import relationship
from sqlalchemy_utils import ChoiceType

from model import Base


class Result(Base):
    ''' Data model for a job result. '''

    __tablename__ = 'result'
    __table_args__ = (
        UniqueConstraint('tracker_id', 'site_url', name='tracker_id_site_url'),
    )

    STATUS_TYPES = [
        (u'f', u'Found'),
        (u'n', u'Not Found'),
        (u'e', u'Error')
    ]

    id = Column(Integer, primary_key=True)
    tracker_id = Column(String(255), nullable=False)
    site_name = Column(String(255), nullable=False)
    site_url = Column(String(255), nullable=False)
    status = Column(ChoiceType(STATUS_TYPES), nullable=False)
    image_file_id = Column(Integer,
                           ForeignKey('file.id',
                                      name='fk_image_file'),
                           nullable=True)
    image_file = relationship('File',
                              lazy='joined',
                              backref='result',
                              uselist=False,
                              cascade='all')
    error = Column(String(255), nullable=True)

    def __init__(self,
                 tracker_id,
                 site_name,
                 site_url,
                 status,
                 image_file_id=None,
                 thumb=None,
                 error=None):
        ''' Constructor. '''

        self.tracker_id = tracker_id
        self.site_name = site_name
        self.site_url = site_url
        self.status = status
        self.image_file_id = image_file_id
        self.thumb = thumb
        self.error = error

    def as_dict(self):
        ''' Return dictionary representation of this result. '''

        return {
            'error': self.error,
            'id': self.id,
            'image_file_id': self.image_file_id,
            'image_file_url': self.image_file.url(),
            'image_file_name': self.image_file.name,
            'site_name': self.site_name,
            'site_url': self.site_url,
            'status': self.status.code,
            'tracker_id': self.tracker_id,
        }

File: static/dart/lib/model/result.dart
/// A model for a username search result.
class Result {

    int id;
    String trackerId;
    String siteName;
    String siteUrl;
    int imageFileId;
    String imageFileUrl;
    int number;
    String status;
    int total;
    String error;

    Result(String trackerId, String siteName, siteUrl, String status, int number, int total) {
        this.status = status;
        this.trackerId = trackerId;
        this.siteName = siteName;
        this.siteUrl = siteUrl;
        this.number = number;
        this.total = total;
    }

    Result.fromJson(Map json) {
        this.status = json['status'];
        this.id = json['id'];
        this.trackerId = json['tracker_id'];
        this.siteName = json['site_name'];
        this.siteUrl = json['site_url'];
        this.imageFileId = json['image_file_id'];
        this.imageFileUrl = json['image_file_url'];
        this.number = json['number'];
        this.total = json['total'];
        this.error = json['error'];
    }
}
